### ISSUE TITLE
Proper HTTP error instead of "Login denied" exception

### ISSUE THREAD

**Issue by HorlogeSkynet** [HomeITAdmin/nextcloud_geoblocker#61]
Hi ! First of all, many thanks for your application.

One proposition/question : what do you think about returning a `403 Forbidden` HTTP error instead of throwing an exception for the blocking feature ?

Such a diff would look like :

```diff
--- a/lib/GeoBlocker/GeoBlocker.php
+++ b/lib/GeoBlocker/GeoBlocker.php
7a8
> use OC\ForbiddenException;
12d12
< use Exception;
92c92
< 							throw new Exception("Login denied!", 1);
---
> 							throw new ForbiddenException("Login denied!");
```

Bye :wave:

> PS : I'm not specifically familiar with Nextcloud code base so this might seem irrelevant.

**Comment by HomeITAdmin** [comment]
I really like it. So the server should send back a http 403 answer when trying to login instead of the general php exception handling? 

I have to do a closer look into the Nextcloud code. I didn't found something obvious on the first glance. 
The ForbiddenException didn't do much different then the Exception in a short test, but making the code and the error log a little bit more descriptive. Also an improvement but not what we are looking for. 

So a hint if and how it is possible in Nextcloud is highly welcome. Another App that is doing something similar would also be helpful.

**Comment by HorlogeSkynet** [comment]
I've just encountered `lib/public/AppFramework/OCS/OCSForbiddenException.php` upstream, maybe something relevant here ? :man_shrugging:

**Comment by HomeITAdmin** [comment]
Thanks. Unfortunately still mainly the same result. 
My current theory is, that if I would have an own route with my own controller, I probably could send back whatever I want including 403. But here I have to hook into the login process and can only use what is given by this hook API. 

What I found is this:
`http_response_code(403); die('Blocked by GEOBLOCKER App.');`

This leads to a 403 answer to the post request for the login with the given text in the answer. At least in my browser just this text is shown afterwards and the login seems to be prevented.
For some reason it is followed by further requests: a 302 and then a 200 with the login screen again, but that is not shown.

So not exactly what I was looking for, but at least closed. And the main advantage would be, that the user get a meaningful answer, why he cannot access Nextcloud.

**Comment by HorlogeSkynet** [comment]
OK, nice try. A bit rough IMHO 🙃
As far as I can tell, I couldn't find a NC "API" to return something specific to the login process when hooking it, maybe we should ask about it to a core developer ?

Anyway, bye 👋

**Comment by HomeITAdmin** [comment]
Makes sense, they should know this. How would you contact the core developer? Last time I tried the forum, which was relative unsucessfull.

**Comment by HorlogeSkynet** [comment]
Absolutely no idea, sorry. I'm _just_ a Nextcloud administrator/user at the moment 🙄 I _guess_ the forum is the way to though...

**Comment by HorlogeSkynet** [comment]
Some more pointers :

* There is a `TwoFactorException` added [here](https://github.com/nextcloud/server/commit/e077e01bf2f2ead37c664e6e1f7339bb3a1d7f85)
* There are a bunch of defined _private_ authentication exceptions [here](https://github.com/nextcloud/server/tree/master/lib/private/Authentication/Exceptions), but not any related to "third-party" blocking decision (as GeoIP in this case)
* I've browsed [twofactor_u2f](https://github.com/nextcloud/twofactor_u2f) application sources, and it actually does not even use `TwoFactorException` 🤔
* Eventually, I've encountered a (public !) "broader" exception [`OCSForbiddenException`](https://github.com/nextcloud/server/blob/905e1918d2796b9a79025283cd6edf2c40f49d77/lib/public/AppFramework/OCS/OCSForbiddenException.php)

Bye !

**Comment by HomeITAdmin** [comment]
Thanks for the pointers. 
Unfortunately also theses exception have the same effect as before. :-(

But I think I found the right one now.
I went back the registration chain for the hook and found this:
```
/**
* Log an user in via login name and password
*
* @param string $uid
* @param string $password
* @return boolean
* @throws LoginException if an app canceled the login process or the user is not enabled
*/
private function loginWithPassword($uid, $password)
```

So this `LoginException  ` sounds exactly as what we need and after testing it has exactly the right behavior also within the hook:
- 403 Answer from the server
- A nice Nextcloud screen telling the customer "Error" and then the error message given in the creation of the Exception
- No error log message caused by the exception, because it is caught by someone who is handling it correctly.
